src/ui/public/agg_types/buckets/date_histogram.js

```diff
@@ -48,12 +50,22 @@ function getBounds(vis) {
 }
 
 function setBounds(agg, force) {
-  if (agg.buckets._alreadySet && !force) return;
-  agg.buckets._alreadySet = true;
-  const bounds = getBounds(agg.vis);
-  agg.buckets.setBounds(agg.fieldIsTimeField() && bounds);
+  const timeBuckets = getTimeBuckets(agg);
+  if (timeBuckets._alreadySet && !force) return;
+  timeBuckets._alreadySet = true;
+  timeBuckets.setBounds(agg.fieldIsTimeField() && getBounds(agg.vis));
 }
 
+function getTimeBuckets(aggConfig) {
```

Comment:
I would find this easier to read if a) we have a single return and b) if we did our configuration of the new bucket before storing it.

```js
function getTimeBuckets(aggConfig) {
  if (!timeBucketCache.has(aggConfig)) {
    const buckets = new TimeBuckets();
    buckets.setInterval(getInterval(aggConfig));
    timeBucketCache.set(aggConfig, buckets);
    setBounds(aggConfig);
  }
  return timeBucketCache.get(aggConfig);
}
```

Reply:
I think a) you will need to discuss with our [style guide](https://github.com/elastic/kibana/blob/master/style_guides/js_style_guide.md#returnthrow-early-from-functions) :D But yeah agree, in some places it just makes it less readable, so can change it here.

b) Unfortunately we need to do that, because the `getInterval` will already try to access buckets again and if we haven't set the cached once at that time we end up in an infinite loop.

Reply:
@timroes are you sure about the `getInterval` here? seems that getInterval is only looking into the aggConfig object and return an object there, and not looking at the timeBucketCache map:
```js
function getInterval(agg) {
  const interval = _.get(agg, ['params', 'interval']);
  if (interval && interval.val === 'custom') {
    return _.get(agg, ['params', 'customInterval']);
  }
  return interval;
}
```